### PR TITLE
feat: using value_schema_manager to reimplement value_generator

### PR DESCRIPTION
**Pull request by levy5307**
### What problem does this PR solve? 
reimplement pegasus_value_generator, by using `value_schema_manager`.
RFC: https://github.com/apache/incubator-pegasus/blob/master/rfcs/2020-10-09-data-version-v3.md

##### Tests 

- Unit test

### CODE REVIEW

File: src/base/value_schema_v0.cpp [hunk 1]
```diff
@@ -19,6 +19,8 @@
 
 #include "value_schema_v0.h"
 
+#include <dsn/utility/endians.h>
```
Comment (Smityz) [inline src/base/value_schema_v0.cpp:22]: Why just add a header here?

File: src/base/value_generator.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+value_generator::generate_value(dsn::string_view user_data, uint32_t expire_ts, uint64_t timetag)
+{
+    value_params params(_write_buf, _write_slices);
+    params.fields[value_field_type::EXPIRE_TIMESTAMP] =
+        dsn::make_unique<expire_timestamp_field>(expire_ts);
+    params.fields[value_field_type::TIME_TAG] = dsn::make_unique<time_tag_field>(timetag);
+    params.fields[value_field_type::USER_DATA] = dsn::make_unique<user_data_field>(user_data);
+
+    auto schema = value_schema_manager::instance().get_latest_value_schema();
+    return schema->generate_value(params);
```
Comment (Smityz) [inline src/base/value_generator.cpp:36]:
```suggestion
    return value_schema_manager::instance().get_latest_value_schema()->generate_value(params);
```
BTW, is there any memory copy happen?

Comment (levy5307) [reply src/base/value_generator.cpp:36]: I think it's ok to use 
```
auto schema = value_schema_manager::instance().get_latest_value_schema();    return schema-&gt;generate_value(params);
```
Your modifications don't make sense.

And there will no memory copy

**Review by hycdong — APPROVED** [review]

File: src/server/rocksdb_wrapper.cpp [hunk 3]
```diff
@@ -119,8 +121,8 @@ int rocksdb_wrapper::write_batch_put_ctx(const db_write_context &ctx,
 
     rocksdb::Slice skey = utils::to_rocksdb_slice(raw_key);
     rocksdb::SliceParts skey_parts(&skey, 1);
-    rocksdb::SliceParts svalue = _value_generator->generate_value(
-        _pegasus_data_version, value, db_expire_ts(expire_sec), new_timetag);
+    rocksdb::SliceParts svalue =
+        _value_generator->generate_value(value, db_expire_ts(expire_sec), new_timetag);
```
[the remark this answers is not in the document]
Comment (levy5307) [reply src/server/rocksdb_wrapper.cpp:125]: I don't want to change the interface in value_generator in this pull request

File: src/base/value_generator.cpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+#include "value_generator.h"
+#include "value_schema_manager.h"
+#include <dsn/utility/smart_pointers.h>
+
+namespace pegasus {
+/// A higher level utility for generating value with given version.
+rocksdb::SliceParts
+value_generator::generate_value(dsn::string_view user_data, uint32_t expire_ts, uint64_t timetag)
+{
+    value_params params(_write_buf, _write_slices);
```
[the remark this answers is not in the document]
Comment (levy5307) [reply src/base/value_generator.cpp:29]: It is used by value_schema::generate_value

File: src/base/value_generator.cpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    params.fields[value_field_type::EXPIRE_TIMESTAMP] =
+        dsn::make_unique<expire_timestamp_field>(expire_ts);
+    params.fields[value_field_type::TIME_TAG] = dsn::make_unique<time_tag_field>(timetag);
+    params.fields[value_field_type::USER_DATA] = dsn::make_unique<user_data_field>(user_data);
+
+    auto schema = value_schema_manager::instance().get_latest_value_schema();
```
[the remark this answers is not in the document]
Comment (levy5307) [reply src/base/value_generator.cpp:35]: We don't want to generate old value if a new version is developped

File: src/base/value_generator.h [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * Helper class for generating value.
+ *
+ * NOTES:
+ * 1. The instance of value_generator must be alive while the returned SliceParts is.
+ * The data of user_data must be alive while the returned SliceParts is, because
+ * we do not copy it. And the returned SliceParts is only valid before the next invoking of
+ * generate_value().
+ * 2. Only the latest version of data is generated in this function, because we want to make
+ * data with older version disappear gradually.
```
[the remark this answers is not in the document]
Comment (levy5307) [reply src/base/value_generator.h:35]: yes

File: src/base/value_generator.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Shuo-Jia) [reply src/base/value_generator.cpp:29]: Oh，I see， here...https://github.com/apache/incubator-pegasus/pull/757/files/e5b729d628b97e7f9243497e2e5b550548b598fe#diff-df7505b9cb2a2157665b192c6f05acd243fc32926cdb5e7727c457a0da1b3f0bR36

**Review by Shuo-Jia — APPROVED** [review]

**Pull request merged**